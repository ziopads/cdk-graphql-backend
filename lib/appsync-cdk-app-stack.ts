import * as cdk from '@aws-cdk/core';
import * as appsync from '@aws-cdk/aws-appsync';
import * as ddb from '@aws-cdk/aws-dynamodb';
import * as lambda from '@aws-cdk/aws-lambda';

export class AppsyncCdkAppStack extends cdk.Stack {

  public readonly api: appsync.GraphqlApi;

  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    this.api = new appsync.GraphqlApi(this, 'Api', {
      name: 'cdk-notes-appsync-api',
      schema: appsync.Schema.fromAsset('graphql/schema.graphql'),
      authorizationConfig: {
        defaultAuthorization: {
          authorizationType: appsync.AuthorizationType.API_KEY,
          apiKeyConfig: {
            expires: cdk.Expiration.after(cdk.Duration.days(365))
          }
        },
      },
      xrayEnabled: true,
    });

    const notesLambda = new lambda.Function(this, 'AppSyncNotesHandler', {
      runtime: lambda.Runtime.NODEJS_12_X,
      handler: 'appsync-ds-main.handler',
      code: lambda.Code.fromAsset('lambda-fns'),
      memorySize: 1024
    });
    
    // set the new Lambda function as a data source for the AppSync API
    const lambdaDs = this.api.addLambdaDataSource('lambdaDatasource', notesLambda);

    // create resolvers to match GraphQL operations in schema
    lambdaDs.createResolver({
      typeName: "Query",
      fieldName: "getNoteById"
    });

    lambdaDs.createResolver({
      typeName: "Query",
      fieldName: "listNotes"
    });

    lambdaDs.createResolver({
      typeName: "Mutation",
      fieldName: "createNote"
    });

    lambdaDs.createResolver({
      typeName: "Mutation",
      fieldName: "deleteNote"
    });

    lambdaDs.createResolver({
      typeName: "Mutation",
      fieldName: "updateNote"
    });

    // create DynamoDB table
    const notesTable = new ddb.Table(this, 'CDKNotesTable', {
      billingMode: ddb.BillingMode.PAY_PER_REQUEST,
      partitionKey: {
        name: 'id',
        type: ddb.AttributeType.STRING,
      },
    });

    // enable the Lambda function to access the DynamoDB table (using IAM)
    notesTable.grantFullAccess(notesLambda)
    
    notesLambda.addEnvironment('NOTES_TABLE', notesTable.tableName);

  }
}
