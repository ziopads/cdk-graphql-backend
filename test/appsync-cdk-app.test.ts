/**
 * Tests AppSyncCdkApp
 * 
 * @group unit
 */
import { SynthUtils } from '@aws-cdk/assert';
import { Stack } from '@aws-cdk/core';
import * as AppSyncCdkApp from '../lib/appsync-cdk-app-stack';
import * as tools from 'graphql-schema-utilities';
import * as graphql from 'graphql';

test('Snapshot test - AppsyncCdkApp', () => {
  const stack = new Stack();
  // WHEN
  new AppSyncCdkApp.AppsyncCdkAppStack(stack, 'MyTestConstruct');
  // THEN
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});

describe('GraphQL Validator', () => {
  describe('#mergeGQLSchemas', () => {
    describe('when loading a schema glob', () => {
      const glob = './graphql/**/*.graphql';
      let schema: graphql.GraphQLSchema;
      beforeAll((done) => {
        tools.mergeGQLSchemas(glob).then((s) => {
          schema = s;
          done();
        });
      });

      it('expect schema to be a graphql schema', (done) => {
        expect(schema).toBeDefined();
        expect(schema).toBeInstanceOf(graphql.GraphQLSchema);
        done();
      });
    });
  });
});