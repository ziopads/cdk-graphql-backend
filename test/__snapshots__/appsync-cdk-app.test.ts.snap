// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test - AppsyncCdkApp 1`] = `Object {}`;
